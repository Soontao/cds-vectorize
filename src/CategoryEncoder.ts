// TODO: handle null value

import { Encoder } from "./Encoder";

export class CategoryEncoder<T = any> implements Encoder<T> {
  #values: any[];

  #valueIndex: Map<T, number>;

  constructor(values: Array<T>) {
    this.#values = values;
    this.#valueIndex = new Map();
    for (const [index, value] of values.entries()) {
      this.#valueIndex.set(value, index);
    }
  }

  encode(value: T): Float32Array {
    const vec = new Float32Array(this.length);
    const index = this.#valueIndex.get(value);
    vec.fill(0);
    if (index) vec[index] = 1;
    return vec;
  }

  decode(vec: Float32Array) {
    return this.#values.at(vec.findIndex((v) => v == 1));
  }

  get length() {
    return this.#values.length;
  }
}

export default CategoryEncoder;
