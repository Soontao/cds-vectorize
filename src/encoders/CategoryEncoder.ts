import { Encoder } from "./Encoder";

export class CategoryEncoder<T = any> implements Encoder<T> {
  #values: any[];

  #valueIndex: Map<T, number>;

  constructor(values: Array<T>) {
    this.#values = values;
    this.#valueIndex = new Map();
    for (const [index, value] of values.entries()) {
      this.#valueIndex.set(value, index);
    }
  }

  encode(value: T): Array<number> {
    if (value == undefined) {
      return new Array(this.length).fill(NaN);
    }
    const vec = new Array(this.length).fill(0);
    const index = this.#valueIndex.get(value);
    if (index !== undefined) vec[index] = 1;
    return vec;
  }

  decode(vec: Array<number>) {
    if (vec.length !== this.length) {
      throw new Error("Invalid vector length");
    }
    const index = vec.findIndex((v) => v === 1);
    if (index < 0) {
      return undefined;
    }
    return this.#values.at(index);
  }

  get length() {
    return this.#values.length;
  }
}

export default CategoryEncoder;
