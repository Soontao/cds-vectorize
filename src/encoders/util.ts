import { Vector } from "./type.js";

export function isNull(value: any): boolean {
  return value === null || value === undefined || (typeof value == "number" && isNaN(value));
}

export function nullVector(length: number): Vector {
  return new Array(length).fill(-(2 ** 16));
}
