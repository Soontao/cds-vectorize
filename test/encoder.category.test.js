// ai
import CategoryEncoder from "../src/encoders/CategoryEncoder.js";
import { nullVector } from "../src/encoders/util.js";

describe("CategoryEncoder Test Suite", () => {
  it("should encode and decode values correctly", () => {
    // Arrange
    const categories = ["A", "B", "C"];
    const encoder = new CategoryEncoder({ values: categories });

    // Act
    const encodedA = encoder.encode("A");
    const encodedB = encoder.encode("B");
    const encodedC = encoder.encode("C");

    // Assert
    expect(encodedA).toEqual([1, 0, 0]);
    expect(encodedB).toEqual([0, 1, 0]);
    expect(encodedC).toEqual([0, 0, 1]);

    // Decode and assert
    expect(encoder.decode(encodedA)).toBe("A");
    expect(encoder.decode(encodedB)).toBe("B");
    expect(encoder.decode(encodedC)).toBe("C");
  });

  it("should support list feature names", () => {
    // Arrange
    const categories = ["A", "B", "C"];
    const encoder = new CategoryEncoder({ name: 'c_1', values: categories });

    expect(encoder.features("c_1")).toMatchSnapshot();
  });

  it("should handle null value", () => {
    // Arrange
    const categories = ["A", "B", "C"];
    const encoder = new CategoryEncoder({ values: categories });

    // Act
    const encodedNull = encoder.encode(null);

    // Assert
    expect(encodedNull).toEqual(nullVector(encoder.length)); // Assuming null is represented as all zeros

    // Decode and assert
    expect(encoder.decode(encodedNull)).toBeNull();
  });

  it("should encode and decode values correctly for different types", () => {
    // Arrange
    const categories = [Symbol("A"), "B", 42, true];
    const encoder = new CategoryEncoder({ values: categories });

    // Act
    const encodedSymbolA = encoder.encode(categories[0]);
    const encodedStringB = encoder.encode("B");
    const encodedNumber42 = encoder.encode(42);
    const encodedBoolTrue = encoder.encode(true);

    // Assert
    expect(encodedSymbolA).toEqual([1, 0, 0, 0]);
    expect(encodedStringB).toEqual([0, 1, 0, 0]);
    expect(encodedNumber42).toEqual([0, 0, 1, 0]);
    expect(encodedBoolTrue).toEqual([0, 0, 0, 1]);

    // Decode and assert
    expect(encoder.decode(encodedSymbolA)).toBe(categories[0]);
    expect(encoder.decode(encodedStringB)).toBe("B");
    expect(encoder.decode(encodedNumber42)).toBe(42);
    expect(encoder.decode(encodedBoolTrue)).toBe(true);
  });
});
