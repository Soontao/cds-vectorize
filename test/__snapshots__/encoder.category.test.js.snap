// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryEncoder Test Suite should support list feature names 1`] = `
[
  "c_1_is_A",
  "c_1_is_B",
  "c_1_is_C",
]
`;
